[DCP][HF] Add option to parallelize reads in HF Storage Reader

ankitageorge opened the pull request:
Parallelize reading of data behind thread_count argument to HFStorageReader
Test plan: ensure existing tests pass and run a job successfully with these changes

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #160205

Differential Revision: [D79478188](https://our.internmc.facebook.com/intern/diff/D79478188/)

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta

ankitageorge committed:
parallelize reads



Differential Revision: [D79478188](https://our.internmc.facebook.com/intern/diff/D79478188/)

[ghstack-poisoned]

ankitageorge committed:
Update on "parallelize reads"


Differential Revision: [D79478188](https://our.internmc.facebook.com/intern/diff/D79478188/)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

ankitageorge committed:
Update on "[DCP][HF] Add option to parallelize reads in HF Storage Reader"

Parallelize reading of data behind thread_count argument to HFStorageReader
Test plan: ensure existing tests pass and run a job successfully with these changes


Differential Revision: [D79478188](https://our.internmc.facebook.com/intern/diff/D79478188/)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

meetv18 reviewed (approved):
LGTM :)

Any way to modularize this and also implement it for `FileSystemReader`?

saumishr commented on torch/distributed/checkpoint/hf_storage.py:
```diff
@@ -201,15 +202,52 @@ class HuggingFaceStorageReader(FileSystemReader):
+        except queue.Empty:
+            pass
 
     def read_data(self, plan: LoadPlan, planner: LoadPlanner) -> Future[None]:
```
Is there a plan to introduce the multithreading for save path as well?

saumishr commented on torch/distributed/checkpoint/hf_storage.py:
```diff
@@ -221,25 +259,47 @@ def read_data(self, plan: LoadPlan, planner: LoadPlanner) -> Future[None]:
-
-                    target_tensor.copy_(tensor)
-                    planner.commit_tensor(req, target_tensor)
+        if self.thread_count <= 1 or len(per_file) <= 1:
```
We need a test  for this method.

ankitageorge committed:
Update on "[DCP][HF] Add option to parallelize reads in HF Storage Reader"

Parallelize reading of data behind thread_count argument to HFStorageReader
Test plan: ensure existing tests pass and run a job successfully with these changes


Differential Revision: [D79478188](https://our.internmc.facebook.com/intern/diff/D79478188/)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

ankitageorge committed:
Update on "[DCP][HF] Add option to parallelize reads in HF Storage Reader"

Parallelize reading of data behind thread_count argument to HFStorageReader
Test plan: ensure existing tests pass and run a job successfully with these changes


Differential Revision: [D79478188](https://our.internmc.facebook.com/intern/diff/D79478188/)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]
